pkg/cachemanager/cachemanager_test/cachemanager_integration_test.go

```diff
@@ -0,0 +1,346 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const (
+	eventuallyTimeout = 10 * time.Second
+	eventuallyTicker  = 2 * time.Second
+)
+
+var cfg *rest.Config
+
+func TestMain(m *testing.M) {
+	testutils.StartControlPlane(m, &cfg, 3)
+}
+
+// TestCacheManager_replay_retries tests that we can retry GVKs that error out in the reply goroutine.
+func TestCacheManager_replay_retries(t *testing.T) {
+	mgr, wm := testutils.SetupManager(t, cfg)
+	c := testclient.NewRetryClient(mgr.GetClient())
+
+	failPlease := make(chan string, 2)
+	reader := fakes.HookReader{
+		Reader: c,
+		ListFunc: func(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+			// Return an error the first go-around.
+			var failKind string
+			select {
+			case failKind = <-failPlease:
+			default:
+			}
+			if failKind != "" && list.GetObjectKind().GroupVersionKind().Kind == failKind {
+				return fmt.Errorf("synthetic failure")
+			}
+			return c.List(ctx, list, opts...)
+		},
+	}
+
+	cacheManager, dataStore, _, ctx := cacheManagerForTest(t, mgr, wm, reader)
+
+	cfClient, ok := dataStore.(*cachemanager.FakeCfClient)
+	require.True(t, ok)
+
+	// seed one gvk
+	configMapGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "ConfigMap"}
+	cm := unstructuredFor(configMapGVK, "config-test-1")
+	require.NoError(t, c.Create(ctx, cm), "creating ConfigMap config-test-1")
+
+	podGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Pod"}
+	pod := unstructuredFor(podGVK, "pod-1")
+	require.NoError(t, c.Create(ctx, pod), "creating Pod pod-1")
+
+	syncSourceOne := aggregator.Key{Source: "source_a", ID: "ID_a"}
+	require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK, podGVK}))
+
+	expected := map[cachemanager.CfDataKey]interface{}{
+		{Gvk: configMapGVK, Key: "default/config-test-1"}: nil,
+		{Gvk: podGVK, Key: "default/pod-1"}:               nil,
+	}
+
+	require.Eventually(t, expectedCheck(cfClient, expected), eventuallyTimeout, eventuallyTicker)
+
+	// set up a scenario where the list from replay will fail a few times
+	failPlease <- "ConfigMapList"
+	failPlease <- "ConfigMapList"
+
+	// this call should schedule a cache wipe and a replay for the configMapGVK
+	require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK}))
+
+	expected2 := map[cachemanager.CfDataKey]interface{}{
+		{Gvk: configMapGVK, Key: "default/config-test-1"}: nil,
+	}
+	require.Eventually(t, expectedCheck(cfClient, expected2), eventuallyTimeout, eventuallyTicker)
+
+	// cleanup
+	require.NoError(t, c.Delete(ctx, cm), "creating ConfigMap config-test-1")
+	require.NoError(t, c.Delete(ctx, pod), "creating ConfigMap pod-1")
+}
+
+// TestCacheManager_AddSourceRemoveSource makes sure that we can add and remove multiple sources
+// and changes to the underlying cache are reflected.
+func TestCacheManager_AddSourceRemoveSource(t *testing.T) {
```

Comment:
We already have test coverage for AddSource and RemoveSource in the other test file.  This test seems to confirm that GVKAggregator is affected by AddSource/RemoveSource AND that manageCache responds to changes in GVKAggregator.  To do this, we get a very big test with many steps.

How much of what is tested here would be confirmed by an e2e test?

Say we were to write an e2e test that:

1. Adds multiple sources
2. Attempts to apply multiple objects that are related to each of the sources and confirms they are denied
3. Removes the sources
4. Successfully applies objects now that those sources are removed (even though the policies are still in place)

We will have to write an e2e test like this _anyways_.  What in this test is not covered by that e2e test?

Reply:
I think these are all good arguments for us to delete this test given that it doesn't add a whole lot of coverage?